File: src/main/views/@default/proxy/waf/ruleUpdate.js
Tea.context(function () {
	this.$delay(function () {
		this.$find("input[name='name']").focus();
		this.sortable();
	});

	this.submitSuccess = function () {
		alert("保存成功");
		window.location = "/proxy/waf/group?wafId=" + this.config.id + "&groupId=" + this.group.id;
	};

	/**
	 * rules
	 */
	this.isAddingRule = false;
	this.ruleIndex = -1;
	this.rules = [];

	if (this.oldRules != null) {
		var that = this;
		this.rules = this.oldRules.$map(function (k, v) {
			return {
				"checkpoint": that.checkpoints.$find(function (k1, v1) {
					return v1.prefix == v.prefix;
				}),
				"param": v.param,
				"operator": that.operators.$find(function (k1, v1) {
					return v1.code == v.operator;
				}),
				"value": v.value
			}
		});
		this.$delay(function () {
			this.refreshTestParams();
		});
	}

	this.testParams = [];
	this.testVisible = false;

	this.addRule = function () {
		this.isAddingRule = true;
		this.ruleIndex = -1;
		this.checkpointPrefix = "";
		this.checkpoint = null;
		this.operatorCode = "match";
		this.changeOperator();
		this.value = "";
	};

	this.cancelAdding = function () {
		this.isAddingRule = false;
		this.ruleIndex = -1;
	};

	this.editRule = function (index) {
		this.ruleIndex = index;
		this.isAddingRule = true;
		this.checkpointPrefix = this.rules[index].checkpoint.prefix;
		this.checkpoint = this.rules[index].checkpoint;
		this.operatorCode = this.rules[index].operator.code;
		this.operator = this.rules[index].operator;
		this.value = this.rules[index].value;
		this.checkpointParam = this.rules[index].param;
	};

	this.confirmAdding = function () {
		if (this.checkpoint == null) {
			alert("请选择一个参数");
			return;
		}
		this.isAddingRule = false;
		if (this.ruleIndex < 0) {
			this.rules.push({
				"checkpoint": this.checkpoint,
				"param": this.checkpointParam,
				"operator": this.operator,
				"value": this.value
			});
		} else {
			this.rules[this.ruleIndex] = {
				"checkpoint": this.checkpoint,
				"param": this.checkpointParam,
				"operator": this.operator,
				"value": this.value
			};
			this.ruleIndex = -1;
		}
		this.refreshTestParams();
		this.isTested = false;
	};

	this.removeRule = function (index) {
		this.rules.$remove(index);
		this.ruleIndex = -1;
		this.isAddingRule = false;
		this.refreshTestParams();
	};

	this.refreshTestParams = function () {
		var testParams = [];
		this.rules.$each(function (k, rule) {
			if (testParams.$find(function (k1, v1) {
				return rule.checkpoint.prefix == v1.prefix && rule.param == v1.param;
			}) != null) {
				return;
			}
			testParams.push({
				"name": rule.checkpoint.name,
				"prefix": rule.checkpoint.prefix,
				"param": rule.param,
				"value": rule.value,
				"description": rule.checkpoint.description
			});
		});
		this.testParams = testParams;
	};

	/**
	 * checkpoint
	 */
	this.checkpointPrefix = "";
	this.checkpoint = null;
	this.checkpointParam = "";
	this.changeCheckpoint = function () {
		if (this.checkpointPrefix.length == 0) {
			this.checkpoint = null;
			return;
		}
		var that = this;
		this.checkpoint = this.checkpoints.$find(function (k, v) {
			return v.prefix == that.checkpointPrefix;
		});
		this.checkpointParam = "";
	};

	/**
	 * operator
	 */
	this.operatorCode = "match";

	this.changeOperator = function () {
		var that = this;
		this.operator = this.operators.$find(function (k, v) {
			return v.code == that.operatorCode;
		});
	};
	this.changeOperator();

	/**
	 * 对比值
	 */
	this.value = "";

	/**
	 * 关系
	 */
	this.connectorValue = this.set.connector;
	this.connectorDescription = "";

	this.changeConnector = function () {
		var that = this;
		this.connectorDescription = this.connectors.$find(function (k, v) {
			return v.value == that.connectorValue;
		}).description;
	};
	this.changeConnector();

	/**
	 * action
	 */
	this.action = this.set.action;

	/**
	 * Test
	 */
	this.matchedIndex = -1;
	this.breakIndex = -1;
	this.isTested = false;

	this.showTestForm = function () {
		this.testVisible = !this.testVisible;
	};

	this.test = function () {
		this.isTested = true;
		this.matchedIndex = -1;
		var form = new FormData(this.$find("#rule-form")[0]);
		form.append("test", "1");
		this.$post("/proxy/waf/group/rule/update")
			.params(form)
			.success(function (resp) {
				this.matchedIndex = resp.data.matchedIndex;
				this.breakIndex = resp.data.breakIndex;
				resp.data.matchLogs.$each(function (k, v) {
					console.log(v);
				});
			});
	};

	/**
	 * 拖动排序
	 */
	this.sortable = function () {
		var box = this.$find(".rules-box")[0];
		var that = this;
		Sortable.create(box, {
			draggable: ".label",
			handle: ".label",
			onStart: function () {

			},
			onUpdate: function (event) {
			}
		});
	};
});

File: internal/teaweb/actions/default/proxy/waf/ruleUpdate.go
package waf

import (
	"encoding/json"
	"github.com/TeaWeb/build/internal/teaconfigs"
	"github.com/TeaWeb/build/internal/teawaf"
	"github.com/TeaWeb/build/internal/teawaf/checkpoints"
	"github.com/TeaWeb/build/internal/teaweb/actions/default/proxy/proxyutils"
	"github.com/TeaWeb/build/internal/teaweb/actions/default/proxy/waf/wafutils"
	"github.com/iwind/TeaGo/actions"
	"github.com/iwind/TeaGo/lists"
	"github.com/iwind/TeaGo/logs"
	"github.com/iwind/TeaGo/maps"
	"regexp"
	"strings"
)

type RuleUpdateAction actions.Action

// 修改规则集
func (this *RuleUpdateAction) RunGet(params struct {
	WafId   string
	GroupId string
	SetId   string
}) {
	waf := teaconfigs.SharedWAFList().FindWAF(params.WafId)
	if waf == nil {
		this.Fail("找不到WAF")
	}

	this.Data["config"] = maps.Map{
		"id":            waf.Id,
		"name":          waf.Name,
		"countInbound":  waf.CountInboundRuleSets(),
		"countOutbound": waf.CountOutboundRuleSets(),
		"inbound":       waf.Inbound,
	}

	group := waf.FindRuleGroup(params.GroupId)
	if group == nil {
		this.Fail("找不到分组")
	}

	this.Data["inbound"] = group.IsInbound
	this.Data["outbound"] = !group.IsInbound

	set := group.FindRuleSet(params.SetId)
	if set == nil {
		this.Fail("找不到规则集")
	}
	err := set.Init()
	if err != nil {
		this.Fail("规则校验失败：" + err.Error())
	}

	reg := regexp.MustCompile("^\\${([\\w.-]+)}$")
	this.Data["set"] = set
	this.Data["oldRules"] = lists.Map(set.Rules, func(k int, v interface{}) interface{} {
		rule := v.(*teawaf.Rule)

		prefix := ""
		param := ""
		result := reg.FindStringSubmatch(rule.Param)
		if len(result) > 0 {
			match := result[1]
			pieces := strings.SplitN(match, ".", 2)
			prefix = pieces[0]
			if len(pieces) == 2 {
				param = pieces[1]
			}
		}

		return maps.Map{
			"prefix":   prefix,
			"param":    param,
			"operator": rule.Operator,
			"value":    rule.Value,
			"case":     rule.IsCaseInsensitive,
			"options":  rule.CheckpointOptions,
		}
	})

	this.Data["group"] = group
	this.Data["connectors"] = []maps.Map{
		{
			"name":        "和 (AND)",
			"value":       teawaf.RuleConnectorAnd,
			"description": "所有规则都满足才视为匹配",
		},
		{
			"name":        "或 (OR)",
			"value":       teawaf.RuleConnectorOr,
			"description": "任一规则满足了就视为匹配",
		},
	}

	// check points
	checkpointList := []maps.Map{}
	for _, def := range checkpoints.AllCheckpoints {
		if (group.IsInbound && def.Instance.IsRequest()) || (!group.IsInbound && !def.Instance.IsRequest()) {
			checkpointList = append(checkpointList, maps.Map{
				"name":         def.Name,
				"prefix":       def.Prefix,
				"description":  def.Description,
				"hasParams":    def.HasParams,
				"paramOptions": def.Instance.ParamOptions(),
				"options": lists.Map(def.Instance.Options(), func(k int, v interface{}) interface{} {
					{
						option, ok := v.(*checkpoints.FieldOption)
						if ok {
							return maps.Map{
								"type":        option.Type(),
								"name":        option.Name,
								"maxLength":   option.MaxLength,
								"code":        option.Code,
								"rightLabel":  option.RightLabel,
								"value":       option.Value,
								"isRequired":  option.IsRequired,
								"size":        option.Size,
								"comment":     option.Comment,
								"placeholder": option.Placeholder,
							}
						}
					}

					{
						option, ok := v.(*checkpoints.OptionsOption)
						if ok {
							return maps.Map{
								"type":       option.Type(),
								"name":       option.Name,
								"code":       option.Code,
								"rightLabel": option.RightLabel,
								"value":      option.Value,
								"isRequired": option.IsRequired,
								"size":       option.Size,
								"comment":    option.Comment,
								"options":    option.Options,
							}
						}
					}

					return maps.Map{}
				}),
			})
		}
	}

	this.Data["checkpoints"] = checkpointList

	this.Data["operators"] = lists.Map(teawaf.AllRuleOperators, func(k int, v interface{}) interface{} {
		def := v.(*teawaf.RuleOperatorDefinition)
		return maps.Map{
			"name":        def.Name,
			"code":        def.Code,
			"description": def.Description,
			"case":        def.CaseInsensitive,
		}
	})

	this.Data["actions"] = lists.Map(teawaf.AllActions, func(k int, v interface{}) interface{} {
		def := v.(*teawaf.ActionDefinition)
		return maps.Map{
			"name":        def.Name,
			"description": def.Description,
			"code":        def.Code,
		}
	})

	this.Show()
}

// 提交测试或者保存
func (this *RuleUpdateAction) RunPost(params struct {
	WafId   string
	GroupId string
	SetId   string

	Name string

	RulePrefixes  []string
	RuleParams    []string
	RuleOperators []string
	RuleValues    []string
	RuleCases     []int
	RuleOptions   []string

	Connector string
	Action    string

	Test         bool
	TestPrefixes []string
	TestParams   []string
	TestValues   []string

	Must *actions.Must
}) {
	// waf
	wafList := teaconfigs.SharedWAFList()
	waf := wafList.FindWAF(params.WafId)
	if waf == nil {
		this.Fail("找不到WAF")
	}

	group := waf.FindRuleGroup(params.GroupId)
	if group == nil {
		this.Fail("找不到Group")
	}

	set := group.FindRuleSet(params.SetId)
	if set == nil {
		this.Fail("找不到规则集")
	}

	set.Name = params.Name
	set.ActionOptions = maps.Map{}
	set.Rules = []*teawaf.Rule{}
	for index, prefix := range params.RulePrefixes {
		if index < len(params.RuleParams) && index < len(params.RuleOperators) && index < len(params.RuleValues) && index < len(params.RuleCases) {
			rule := teawaf.NewRule()
			rule.Operator = params.RuleOperators[index]

			param := params.RuleParams[index]
			if len(param) > 0 {
				rule.Param = "${" + prefix + "." + param + "}"
			} else {
				rule.Param = "${" + prefix + "}"
			}
			rule.Value = params.RuleValues[index]
			rule.IsCaseInsensitive = params.RuleCases[index] == 1

			// 选项
			options := params.RuleOptions[index]
			if len(options) > 0 {
				arr := []maps.Map{}
				err := json.Unmarshal([]byte(options), &arr)
				if err != nil {
					logs.Error(err)
				} else {
					rule.CheckpointOptions = map[string]string{}
					for _, m := range arr {
						code := m.GetString("code")
						value := m.GetString("value")
						rule.CheckpointOptions[code] = value
					}
				}
			}

			// 校验
			err := rule.Init()
			if err != nil {
				this.Fail("校验规则 '" + rule.Param + " " + rule.Operator + " " + rule.Value + "' 失败，原因：" + err.Error())
			}

			set.AddRule(rule)
		}
	}
	set.Connector = params.Connector

	// action
	set.Action = params.Action
	set.ActionOptions = maps.Map{}
	for k, v := range this.ParamsMap {
		if len(v) == 0 {
			continue
		}
		index := strings.Index(k, "action_")
		if index > -1 {
			set.ActionOptions[k[len("action_"):]] = v[0]
		}
	}

	// 测试
	if params.Test {
		err := set.Init()
		if err != nil {
			this.Fail("校验错误：" + err.Error())
		}

		matchedIndex := -1
		breakIndex := -1
		matchLogs := []string{"start matching ...", "==="}
	Loop:
		for index, prefix := range params.TestPrefixes {
			if index < len(params.TestParams) && index < len(params.TestValues) {
				param := ""
				if len(params.TestParams[index]) == 0 {
					param = "${" + prefix + "}"
				} else {
					param = "${" + prefix + "." + params.TestParams[index] + "}"
				}

				breakIndex = index

				for _, rule := range set.Rules {
					if rule.Param == param {
						value := params.TestValues[index]
						if rule.Test(value) {
							matchLogs = append(matchLogs, "rule: "+rule.Param+" "+rule.Operator+" "+rule.Value+"\ncompare: "+value+"\nresult:true")

							if set.Connector == teawaf.RuleConnectorOr {
								matchedIndex = index
								break Loop
							}

							if set.Connector == teawaf.RuleConnectorAnd {
								matchedIndex = index
							}
						} else {
							matchLogs = append(matchLogs, "rule: "+rule.Param+" "+rule.Operator+" "+rule.Value+"\ncompare: "+value+"\nresult:false")

							if set.Connector == teawaf.RuleConnectorAnd {
								matchedIndex = -1
								break Loop
							}
						}
					}
				}
			}
		}

		this.Data["matchedIndex"] = matchedIndex
		this.Data["breakIndex"] = breakIndex
		this.Data["matchLogs"] = matchLogs
		this.Success()
	}

	// 保存
	params.Must.
		Field("name", params.Name).
		Require("请输入规则集名称")

	err := wafList.SaveWAF(waf)
	if err != nil {
		this.Fail("保存失败：" + err.Error())
	}

	// 通知刷新
	if wafutils.IsPolicyUsed(waf.Id) {
		proxyutils.NotifyChange()
	}

	this.Success()
}
